cantools/database/utils.py

```diff
@@ -60,16 +59,17 @@ def _encode_fields(fields: Sequence[Union["Signal", "Data"]],
         value = data[field.name]
 
         if isinstance(value, (float, int)):
+            dtype = float if field.is_float else int
             if scaling:
-                if field.is_float:
-                    unpacked[field.name] = (value - field.offset) / field.scale
-                    continue
-                else:
-                    decimal = (Decimal(value) - Decimal(field.offset)) / Decimal(field.scale)
-                    unpacked[field.name] = int(decimal.to_integral_value())
+                if field.offset == 0 and field.scale == 1:
+                    # treat special case to avoid introduction of unnecessary rounding error
+                    unpacked[field.name] = dtype(value)
                     continue
 
-            unpacked[field.name] = value
+                unpacked[field.name] = dtype((value - field.offset) / field.scale)
```

Comment:
I'm wondering: if `dtype` is `int`,  should this be rounded instead of truncated? (I'm eternally confused by the signal scaling code...)

Reply:
You're right, fixed.